contracts/pcv/curve/CurveSwapper.sol

```diff
@@ -0,0 +1,195 @@
+import "../IPCVSwapper.sol";
+import "../../Constants.sol";
+import "../../refs/CoreRef.sol";
+import "../../core/TribeRoles.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+/// @title CurveSwapper
+/// @author eswak
+/// @notice a contract that performs swaps on Curve.
+/// @dev this implementation assumes a StableSwap metapool, where `tokenSpent` or
+/// `tokenReceived` is the `curvePool.coins(0)` or `curvePool.coins(1)`, and the
+/// other is a Curve BasePool that contains the second token.
+contract CurveSwapper is CoreRef, IPCVSwapper {
+    using SafeERC20 for IERC20;
+
+    address private constant ERC20_3CRV = 0x6c3F90f043a72FA612cbac8115EE7e52BDe6E490;
+    address private constant CURVE_3POOL = 0xbEbc44782C7dB0a1A60Cb6fe97d0b483032FF1C7;
+
+    // ------------- Events -------------
+
+    event WithdrawERC20(address indexed _caller, address indexed _token, address indexed _to, uint256 _amount);
+    event UpdateMaximumSlippage(uint256 maximumSlippageBasisPoints);
+
+    // -------------- State -------------
+    /// @notice the Curve pool used for swapping
+    address public immutable curvePool;
+    /// @notice the index of `tokenSpent` in `curvePool` for underlying swap
+    int128 public immutable i;
+    /// @notice the index of `tokenReceived` in `curvePool` for underlying swap
+    int128 public immutable j;
+
+    /// @notice the token to sell
+    address public immutable override tokenSpent;
+
+    /// @notice the token to buy
+    address public immutable override tokenReceived;
+
+    /// @notice the address to send `tokenReceived`
+    address public override tokenReceivingAddress;
+
+    // @notice Maximum tolerated slippage for swaps in basis points (10_000)
+    uint256 public maxSlippageBps;
+
+    /**
+    @notice constructor for CurveSwapper
+    @param _core Core contract to reference
+    @param _curvePool the Curve pool used to swap
+    @param _tokenSpent the token to be auctioned
+    @param _tokenReceived the token to buy
+    @param _tokenReceivingAddress the address to send `tokenReceived`
+    @param _maxSlippageBps the maximum tolerated slippage for swaps in basis points (10_000)
+     */
+    constructor(
+        address _core,
+        address _curvePool,
+        address _tokenSpent,
+        address _tokenReceived,
+        address _tokenReceivingAddress,
+        uint256 _maxSlippageBps
+    ) CoreRef(_core) {
+        curvePool = _curvePool;
+        tokenSpent = _tokenSpent;
+        tokenReceived = _tokenReceived;
+        _setReceivingAddress(_tokenReceivingAddress);
+        _setMaximumSlippage(_maxSlippageBps);
+
+        // find the i and j index for underlying swap, where
+        // i = _tokenSpent, and j = _tokenReceived
+        address coins0 = ICurvePool(_curvePool).coins(0);
+        address coins1 = ICurvePool(_curvePool).coins(1);
+        address _basePool;
+        int128 _i;
+        int128 _j;
+        if (_tokenSpent == coins0) {
```

Comment:
Not a fan of this logic here, it's hard to follow and feels a spot for mistakes. I'm not sure what would be an improvement though - is there a well tested Curve library we can import that will sort for us?

Overall, I like the constructor API of just being able to pass a `tokenSpent` and `tokenReceived` and all index calculations are abstracted away; but wondering if there's a cleaner/nicer way of getting them

Reply:
I'm not a fan either and honestly we could pass `i` and `j` as constructor args, they won't change

Reply:
passed i & j as constructor params

Reply:
Nice